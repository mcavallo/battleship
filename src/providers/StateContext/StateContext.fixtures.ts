import cloneDeep from 'lodash/cloneDeep';

import { Board, Ship, ShipTemplate, ShipsMap } from './StateContext.types.ts';

export const SHIP_ONE_TEMPLATE: ShipTemplate = { name: 'one', hits: 0, size: 4 };

export const SHIP_ONE: Ship = { ...SHIP_ONE_TEMPLATE, id: 1 };

export const SHIP_TWO_TEMPLATE: ShipTemplate = { name: 'two', hits: 2, size: 4 };

export const SHIP_TWO: Ship = { ...SHIP_TWO_TEMPLATE, id: 2 };

export const SHIP_THREE_TEMPLATE: ShipTemplate = { name: 'three', hits: 4, size: 4 };

export const SHIP_THREE: Ship = { ...SHIP_THREE_TEMPLATE, id: 3 };

export const SHIP_FOUR_TEMPLATE: ShipTemplate = { name: 'four', hits: 0, size: 2 };

export const SHIP_FOUR: Ship = { ...SHIP_FOUR_TEMPLATE, id: 4 };

export const SHIP_FIVE_TEMPLATE: ShipTemplate = { name: 'five', hits: 0, size: 3 };

export const SHIP_FIVE: Ship = { ...SHIP_FIVE_TEMPLATE, id: 5 };

export const SHIP_SIX_TEMPLATE: ShipTemplate = { name: 'six', hits: 0, size: 4 };

export const SHIP_SIX: Ship = { ...SHIP_SIX_TEMPLATE, id: 6 };

export const SHIP_MAP_ONE: ShipsMap = {
  1: SHIP_ONE,
};

export const SHIP_MAP_ONE_TWO: ShipsMap = {
  1: SHIP_ONE,
  2: SHIP_TWO,
};

export const SHIP_MAP_ONE_TWO_THREE: ShipsMap = {
  1: SHIP_ONE,
  2: SHIP_TWO,
  3: SHIP_THREE,
};

export const BOARD_SIZE_2: Board = [
  [0, 0],
  [0, 0],
];
export const BOARD_SIZE_5: Board = [
  [0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0],
];

export const BOARD_SIZE_10: Board = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];

/**
 * Fills a row with sequential numbers.
 */
export function fillRow(board: Board, y: number) {
  const clone = cloneDeep(board);
  clone[y] = clone[y].map((_, idx) => idx + 1);
  return clone;
}

/**
 * Fills a column with sequential numbers.
 */
export function fillCol(board: Board, x: number) {
  const clone = cloneDeep(board);

  for (let i = 0; i < clone.length; i++) {
    clone[i][x] = i + 1;
  }

  return clone;
}
