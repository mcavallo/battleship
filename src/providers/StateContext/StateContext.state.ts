import { BOARD_SIZE } from '@/constants';

import {
  EventKind,
  GameAction,
  GameActionKind,
  GameState,
  ShipTemplate,
} from './StateContext.types';
import {
  calculateScore,
  computeAttack,
  computeGameStats,
  computeShipsPlacement,
  generateEmptyBoard,
  generateShipsMaps,
} from './StateContext.utils.ts';

export const BATTLESHIP_TEMPLATE: ShipTemplate = {
  name: 'Battleship',
  hits: 0,
  size: 5,
} as const;

export const DESTROYER_TEMPLATE: ShipTemplate = {
  name: 'Destroyer',
  hits: 0,
  size: 4,
} as const;

export function getInitialState(): GameState {
  const shipsMap = generateShipsMaps([BATTLESHIP_TEMPLATE, DESTROYER_TEMPLATE, DESTROYER_TEMPLATE]);

  return {
    attacksMap: {},
    board: generateEmptyBoard(BOARD_SIZE),
    hasGameEnded: false,
    hasGameStarted: false,
    isDebug: false,
    log: [],
    shipsMap,
    score: calculateScore(shipsMap),
  } as const;
}

export function gameReducer(state: GameState, action: GameAction) {
  // @ts-expect-error dont mind this here!
  console.log(`@[${action.type}]`, action.payload!);

  switch (action.type) {
    case GameActionKind.Attack: {
      const newState = structuredClone(state);
      const { ts, coords } = action.payload;

      const computedState = {
        ...newState,
        ...computeAttack(newState, ts, coords),
      };

      if (computedState.score.current === computedState.score.win) {
        return gameReducer(computedState, {
          type: GameActionKind.End,
          payload: {
            ts,
          },
        });
      }

      return computedState;
    }

    case GameActionKind.End: {
      const newState = structuredClone(state);
      const { log } = newState;
      const { ts } = action.payload;

      return {
        ...newState,
        hasGameEnded: true,
        log: log.concat({
          ts,
          event: {
            stats: computeGameStats(newState),
            type: EventKind.End,
          },
        }),
      };
    }

    case GameActionKind.Start: {
      const newState = structuredClone(getInitialState());
      const { board, shipsMap, log } = newState;
      const { ts } = action.payload;

      return {
        ...newState,
        board: computeShipsPlacement(ts, board, shipsMap),
        hasGameStarted: true,
        log: log.concat({
          ts,
          event: {
            type: EventKind.Start,
          },
        }),
        isDebug: state.isDebug,
      };
    }

    case GameActionKind.ToggleDebug: {
      return {
        ...state,
        isDebug: !state.isDebug,
      };
    }

    default:
      return state;
  }
}
