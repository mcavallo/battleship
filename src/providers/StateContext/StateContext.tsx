import React, { createContext, useContext, useMemo, useReducer } from 'react';

import { gameReducer, getInitialState } from './StateContext.state.ts';
import { GameAction, GameState } from './StateContext.types.ts';

export interface StateContextValues {
  state: GameState;
  dispatch: React.Dispatch<GameAction>;
}

interface StateContextProviderProps {
  children: React.ReactNode;
}

export const StateContext = createContext<StateContextValues>({} as StateContextValues);

// eslint-disable-next-line react-refresh/only-export-components
export const useGameState = () => {
  const value = useContext(StateContext);

  if (value === null) {
    throw new Error(`'useGameState' should only be used inside 'StateContext'.`);
  }

  return value;
};

export const StateContextProvider = ({ children }: StateContextProviderProps) => {
  const [state, dispatch] = useReducer(gameReducer, getInitialState());
  const value = useMemo(() => ({ state, dispatch }), [state, dispatch]);

  return <StateContext.Provider value={value}>{children}</StateContext.Provider>;
};
