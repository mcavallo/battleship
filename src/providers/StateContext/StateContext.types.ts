export enum Direction {
  Horizontal = 0,
  Vertical = 1,
}

export type Coords = {
  x: number;
  y: number;
};

export type Board = number[][];

export type ShipsMap = Record<number, Ship>;

export type ShipTemplate = {
  name: string;
  size: number;
  hits: number;
};

export type Ship = ShipTemplate & {
  id: number;
};

export type Placement = Coords & {
  direction: Direction;
  size: number;
};

export type GameScore = {
  current: number;
  win: number;
};

export type GameStats = {
  attacks: number;
  hitRatio: number;
  hits: number;
  misses: number;
  score: number;
  sunk: number;
};

export enum EventKind {
  End = 'End',
  Hit = 'Hit',
  Miss = 'Miss',
  Sink = 'Sink',
  Start = 'Start',
}

export type EndEvent = {
  stats: GameStats;
  type: EventKind.End;
};

export type HitEvent = Coords & {
  targetId: number;
  type: EventKind.Hit;
};

export type MissEvent = Coords & {
  type: EventKind.Miss;
};

export type SinkEvent = Coords & {
  targetId: number;
  type: EventKind.Sink;
};

export type StartEvent = {
  type: EventKind.Start;
};

export type Event = EndEvent | HitEvent | MissEvent | SinkEvent | StartEvent;

export type LogEntry = {
  ts: number;
  event: Event;
};

export type GameState = {
  attacksMap: Record<string, boolean>;
  board: Board;
  hasGameEnded: boolean;
  hasGameStarted: boolean;
  isDebug: boolean;
  log: LogEntry[];
  shipsMap: ShipsMap;
  score: GameScore;
};

export enum GameActionKind {
  Attack = 'Attack',
  End = 'End',
  Introspection = 'Introspection', // Used only for testing
  Start = 'Start',
  ToggleDebug = 'ToggleDebug',
}

export interface AttackAction {
  type: GameActionKind.Attack;
  payload: {
    coords: Coords;
    ts: number;
  };
}

export interface EndAction {
  type: GameActionKind.End;
  payload: {
    ts: number;
  };
}

export interface IntrospectionAction {
  type: GameActionKind.Introspection;
}

export interface StartAction {
  type: GameActionKind.Start;
  payload: {
    ts: number;
  };
}

export interface ToggleDebugAction {
  type: GameActionKind.ToggleDebug;
}

export type GameAction =
  | AttackAction
  | EndAction
  | IntrospectionAction
  | StartAction
  | ToggleDebugAction;
