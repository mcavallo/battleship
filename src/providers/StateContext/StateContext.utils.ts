import cloneDeep from 'lodash/cloneDeep';
import Rand from 'rand-seed';

import { LOOP_PROTECTION_LIMIT } from '@/constants';

import {
  Board,
  Coords,
  Direction,
  EventKind,
  GameScore,
  GameState,
  GameStats,
  Placement,
  ShipTemplate,
  ShipsMap,
} from './StateContext.types';

/*
 * Generates a seeded random number between two numbers.
 */
export function randomInt(rand: Rand, min: number, max: number) {
  return Math.floor(rand.next() * (max - min + 1) + min);
}

/*
 * Generates an empty game board of the given size.
 */
export function generateEmptyBoard(size: number): Board {
  const board: Board = [];
  if (size < 2) {
    throw new Error('Invalid board size.');
  }

  for (let i = 0; i < size; i++) {
    board.push([]);
    for (let j = 0; j < size; j++) {
      board[i].push(0);
    }
  }

  return board;
}

/*
 * Calculates the score for the game based on the amount and size of the ships and
 * the amount of hits.
 */
export function calculateScore(shipsMap: ShipsMap): GameScore {
  return Object.values(shipsMap).reduce(
    (acc, ship) => {
      acc.current += ship.hits;
      acc.win += ship.size;
      return acc;
    },
    { current: 0, win: 0 },
  );
}

/*
 * Calculates the required win score based on the amount and size of the ships.
 */
export function generateShipsMaps(ships: ShipTemplate[]): ShipsMap {
  let currentId = 1;
  return ships.reduce<ShipsMap>((acc, ship) => {
    acc[currentId] = { ...ship, id: currentId };
    currentId++;
    return acc;
  }, {});
}

/*
 * Given a set of coordinates and a size, it returns a rows slice of the board.
 */
export function getRowsSlice(board: Board, { x, y, size }: { x: number; y: number; size: number }) {
  if (board[y].length < x + size) {
    throw new Error(`getRowsSlice on ${x}:${y} with size ${size} is out of bounds.`);
  }

  const rows: number[] = [];

  for (let i = 0; i < size; i++) {
    rows.push(board[y][x + i]);
  }

  return rows;
}

/*
 * Given a set of coordinates and a size, it returns a columns slice of the board.
 */
export function getColsSlice(board: Board, { x, y, size }: { x: number; y: number; size: number }) {
  if (board.length < y + size) {
    throw new Error(`getColsSlice on ${x}:${y} with size ${size} is out of bounds.`);
  }

  const cols: number[] = [];

  for (let i = 0; i < size; i++) {
    cols.push(board[y + i][x]);
  }

  return cols;
}

/*
 * Returns a valid placement if a ship can be placed in the desired coordinates,
 * otherwise it returns null.
 */
export function findValidPlacement(board: Board, placement: Placement): Placement | null {
  const { direction, size, x, y } = placement;

  if (direction === Direction.Horizontal) {
    if (x + size > board[y].length) {
      return null;
    }

    const rows = getRowsSlice(board, { x, y, size });
    return rows.some((v) => v !== 0) ? null : placement;
  } else {
    if (y + size > board.length) {
      return null;
    }

    const cols = getColsSlice(board, { x, y, size });
    return cols.some((v) => v !== 0) ? null : placement;
  }
}

/*
 * Returns a string representation of a placement.
 */
export function placementAsString(placement: Placement) {
  const direction = placement.direction === Direction.Horizontal ? 'Horizontal' : 'Vertical';
  return `coords: ${placement.x}:${placement.y}, direction: ${direction}, size: ${placement.size}`;
}

/*
 * Places a ship on the board.
 */
export function placeShip(board: Board, shipId: number, placement: Placement) {
  const verifiedPlacement = findValidPlacement(board, placement);

  if (verifiedPlacement === null) {
    throw new Error(`Invalid placement at ${placementAsString(placement)}.`);
  }

  const newBoard = cloneDeep(board);

  if (placement.direction === Direction.Horizontal) {
    for (let i = placement.x; i < placement.x + placement.size; i++) {
      newBoard[placement.y][i] = shipId;
    }
  } else {
    for (let i = placement.y; i < placement.y + placement.size; i++) {
      newBoard[i][placement.x] = shipId;
    }
  }

  return newBoard;
}

/*
 * Places all ships on the board.
 */
export function computeShipsPlacement(ts: number, board: Board, shipsMap: ShipsMap) {
  let newBoard = cloneDeep(board);
  const rand = new Rand(String(ts));
  const shipIds = Object.values(shipsMap).map((ship) => ship.id);
  let shipId = shipIds.shift();
  let loopCount = 0;

  while (shipId) {
    if (loopCount > LOOP_PROTECTION_LIMIT) {
      throw new Error(`Ship placement is infinite looping.`);
    }

    const x = randomInt(rand, 0, board.length - 1);
    const y = randomInt(rand, 0, board.length - 1);
    const direction = randomInt(rand, 0, 1);
    const ship = shipsMap[shipId];
    const placement = { direction, size: ship.size, x, y };

    try {
      newBoard = placeShip(newBoard, ship.id, placement);
      shipId = shipIds.shift();
    } catch {
      // Errors are expected in this context
    }

    loopCount++;
  }

  return newBoard;
}

/*
 * Computes an attack at the given coordinates.
 */
export function computeAttack(state: GameState, ts: number, coords: Coords) {
  const { attacksMap, board, log, score, shipsMap } = structuredClone(state);
  const [shipId] = getRowsSlice(board, { ...coords, size: 1 });
  const isHit = shipId !== 0;

  attacksMap[`${coords.y}${coords.x}`] = isHit;

  if (!isHit) {
    log.push({
      ts,
      event: {
        ...coords,
        type: EventKind.Miss,
      },
    });

    return {
      attacksMap,
      log,
      score,
    };
  }

  const target = shipsMap[shipId];
  const eventData = {
    ...coords,
    targetId: target.id,
  };

  target.hits++;

  log.push({
    ts,
    event: {
      ...eventData,
      type: EventKind.Hit,
    },
  });

  if (target.hits >= target.size) {
    log.push({
      ts,
      event: {
        ...eventData,
        type: EventKind.Sink,
      },
    });
  }

  return {
    attacksMap,
    log,
    score: calculateScore(shipsMap),
    shipsMap,
  };
}

/*
 * Aggregates the game stats.
 */
export function computeGameStats(state: GameState): GameStats {
  const stats = state.log.reduce(
    (acc, entry) => {
      switch (entry.event.type) {
        case EventKind.Miss:
          acc.attacks++;
          acc.misses++;
          break;
        case EventKind.Hit:
          acc.attacks++;
          acc.hits++;
          break;
        case EventKind.Sink:
          acc.sunk++;
          break;
        default:
          break;
      }

      return acc;
    },
    { attacks: 0, hitRatio: 0, hits: 0, misses: 0, score: 0, sunk: 0 },
  );

  stats.hitRatio = stats.attacks > 0 ? (stats.hits / (stats.hits + stats.misses)) * 100 : 0;
  stats.score = state.score.current;

  return stats;
}

/*
 * Checks if the victory condition for the game has been reached.
 */
export function isVictoryConditionReached(state: GameState) {
  return state.score.current === state.score.win;
}
