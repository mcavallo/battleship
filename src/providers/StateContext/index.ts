export { useGameState, StateContextProvider } from './StateContext';
export { GameActionKind } from './StateContext.types';
