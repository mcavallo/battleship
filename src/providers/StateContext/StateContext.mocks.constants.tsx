import { GameAction, GameActionKind, GameState } from './StateContext.types';

export const DEFAULT_MOCKED_STATE: GameState = {
  attacksMap: {},
  board: [],
  hasGameEnded: false,
  hasGameStarted: false,
  isDebug: false,
  log: [],
  shipsMap: {},
  score: {
    current: 0,
    win: 0,
  },
};

export const INTROSPECTION_ACTION: GameAction = {
  type: GameActionKind.Introspection,
};

export const INTROSPECTION_STATE: GameState = DEFAULT_MOCKED_STATE;
