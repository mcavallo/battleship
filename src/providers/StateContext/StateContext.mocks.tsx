import merge from 'lodash/merge';
import React, { Dispatch } from 'react';

import { GameAction, GameState } from '@/providers/StateContext/StateContext.types.ts';

import { StateContext } from './StateContext';

export type PartialStateContextProviderValue = {
  state?: Partial<GameState>;
  dispatch?: Dispatch<GameAction>;
};

interface MockedStateContextProviderProps {
  children?: React.ReactNode;
  value?: PartialStateContextProviderValue;
}

const defaultState = {
  attacksMap: {},
  board: [],
  hasGameEnded: false,
  hasGameStarted: false,
  isDebug: false,
  log: [],
  shipsMap: {},
  score: {
    current: 0,
    win: 0,
  },
};

export const MockedStateContextProvider = ({
  children,
  value,
}: MockedStateContextProviderProps) => {
  const valueWithDefaults = {
    state: merge(Object.assign({}, defaultState), value?.state || {}),
    dispatch: value?.dispatch || (() => {}),
  };

  return <StateContext.Provider value={valueWithDefaults}>{children}</StateContext.Provider>;
};
