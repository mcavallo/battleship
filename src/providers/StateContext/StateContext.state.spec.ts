import MockDate from 'mockdate';

import { gameReducer, getInitialState } from './StateContext.state';
import { GameAction, GameActionKind } from './StateContext.types';
import { calculateScore } from './StateContext.utils';

const NOW = '2024-05-24T00:00:00.000Z';

jest.mock('./StateContext.utils', () => {
  const original = jest.requireActual('./StateContext.utils');
  return {
    __esModule: true,
    ...original,
    calculateScore: jest.fn().mockImplementation(original.calculateScore),
  };
});

afterAll(() => {
  MockDate.reset();
});

beforeEach(() => {
  MockDate.set(NOW);
});

describe(`getInitialState`, () => {
  it(`returns the default initial state`, () => {
    expect(getInitialState()).toMatchSnapshot();
  });
});

describe(`gameReducer`, () => {
  describe(`fallback`, () => {
    it('returns the state unchanged for unhandled actions', () => {
      const initialState = getInitialState();
      const unknownAction = {
        type: 'UNKNOWN',
      };
      const result = gameReducer(initialState, unknownAction as GameAction);

      expect(result).toEqual(initialState);
    });
  });

  describe(`AttackAction`, () => {
    it(`computes the new state correctly`, () => {
      const initialState = getInitialState();
      const result = gameReducer(initialState, {
        type: GameActionKind.Attack,
        payload: {
          ts: Date.now(),
          coords: {
            x: 0,
            y: 0,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it(`skips the attack if the cell was already attacked`, () => {
      const initialState = getInitialState();
      initialState.attacksMap['00'] = true;

      let result = gameReducer(initialState, {
        type: GameActionKind.Attack,
        payload: {
          ts: Date.now(),
          coords: {
            x: 0,
            y: 0,
          },
        },
      });

      expect(result).toEqual(initialState);

      initialState.attacksMap['00'] = false;

      result = gameReducer(initialState, {
        type: GameActionKind.Attack,
        payload: {
          ts: Date.now(),
          coords: {
            x: 0,
            y: 0,
          },
        },
      });

      expect(result).toEqual(initialState);
    });

    it('runs the EndAction when the win condition is met', () => {
      (calculateScore as jest.Mock).mockReturnValueOnce({
        current: 13,
        win: 13,
      });

      const initialState = getInitialState();
      const result = gameReducer(initialState, {
        type: GameActionKind.Attack,
        payload: {
          ts: Date.now(),
          coords: {
            x: 0,
            y: 0,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });
  });

  describe(`EndAction`, () => {
    it(`computes the new state correctly`, () => {
      const initialState = getInitialState();
      const result = gameReducer(initialState, {
        type: GameActionKind.End,
        payload: {
          ts: Date.now(),
        },
      });

      expect(result).toMatchSnapshot();
    });
  });

  describe(`StartAction`, () => {
    it(`computes the new state correctly`, () => {
      const initialState = getInitialState();
      const result = gameReducer(initialState, {
        type: GameActionKind.Start,
        payload: {
          ts: Date.now(),
        },
      });

      expect(result).toMatchSnapshot();
    });

    it(`preserves the isDebug value`, () => {
      const initialState = getInitialState();
      expect(initialState.isDebug).toEqual(false);

      initialState.isDebug = true;

      const result = gameReducer(initialState, {
        type: GameActionKind.Start,
        payload: {
          ts: Date.now(),
        },
      });

      expect(initialState.isDebug).toEqual(true);
      expect(result.isDebug).toEqual(true);
    });
  });

  describe(`ToggleDebugAction`, () => {
    it(`computes the new state correctly`, () => {
      const initialState = getInitialState();
      const result = gameReducer(initialState, {
        type: GameActionKind.ToggleDebug,
      });

      expect(initialState.isDebug).toEqual(false);
      expect(result).toMatchSnapshot();
    });
  });
});
