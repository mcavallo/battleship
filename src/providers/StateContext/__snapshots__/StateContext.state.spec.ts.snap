// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameReducer AttackAction computes the new state correctly 1`] = `
{
  "attacksMap": {
    "00": false,
  },
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": false,
  "hasGameStarted": false,
  "isDebug": false,
  "log": [
    {
      "event": {
        "type": "Miss",
        "x": 0,
        "y": 0,
      },
      "ts": 1716508800000,
    },
  ],
  "score": {
    "current": 0,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;

exports[`gameReducer AttackAction runs the EndAction when the win condition is met 1`] = `
{
  "attacksMap": {
    "00": false,
  },
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": true,
  "hasGameStarted": false,
  "isDebug": false,
  "log": [
    {
      "event": {
        "type": "Miss",
        "x": 0,
        "y": 0,
      },
      "ts": 1716508800000,
    },
    {
      "event": {
        "stats": {
          "attacks": 1,
          "hitRatio": 0,
          "hits": 0,
          "misses": 1,
          "score": 13,
          "sunk": 0,
        },
        "type": "End",
      },
      "ts": 1716508800000,
    },
  ],
  "score": {
    "current": 13,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;

exports[`gameReducer EndAction computes the new state correctly 1`] = `
{
  "attacksMap": {},
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": true,
  "hasGameStarted": false,
  "isDebug": false,
  "log": [
    {
      "event": {
        "stats": {
          "attacks": 0,
          "hitRatio": 0,
          "hits": 0,
          "misses": 0,
          "score": 0,
          "sunk": 0,
        },
        "type": "End",
      },
      "ts": 1716508800000,
    },
  ],
  "score": {
    "current": 0,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;

exports[`gameReducer StartAction computes the new state correctly 1`] = `
{
  "attacksMap": {},
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      2,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      2,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      2,
      0,
      1,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      2,
      0,
      1,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
    ],
    [
      0,
      3,
      3,
      3,
      3,
      0,
      1,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": false,
  "hasGameStarted": true,
  "isDebug": false,
  "log": [
    {
      "event": {
        "type": "Start",
      },
      "ts": 1716508800000,
    },
  ],
  "score": {
    "current": 0,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;

exports[`gameReducer ToggleDebugAction computes the new state correctly 1`] = `
{
  "attacksMap": {},
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": false,
  "hasGameStarted": false,
  "isDebug": true,
  "log": [],
  "score": {
    "current": 0,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;

exports[`getInitialState returns the default initial state 1`] = `
{
  "attacksMap": {},
  "board": [
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "hasGameEnded": false,
  "hasGameStarted": false,
  "isDebug": false,
  "log": [],
  "score": {
    "current": 0,
    "win": 13,
  },
  "shipsMap": {
    "1": {
      "hits": 0,
      "id": 1,
      "name": "Battleship",
      "size": 5,
    },
    "2": {
      "hits": 0,
      "id": 2,
      "name": "Destroyer",
      "size": 4,
    },
    "3": {
      "hits": 0,
      "id": 3,
      "name": "Destroyer",
      "size": 4,
    },
  },
}
`;
